pyradiosky/skymodel.py

```diff
@@ -905,41 +905,52 @@ def skymodel_to_array(sky):
     catalog_table : recarray
         recarray equivalent to SkyModel data.
 
+    Notes
+    -----
+    This stores all SkyModel data in a contiguous array
+    that can be more easily handled with numpy.
+    This is used by pyuvsim for sharing catalog data via MPI.
     """
     sky.check()
     max_name_len = np.max([len(name) for name in sky.name])
-    fieldtypes = ["U" + str(max_name_len), "f8", "f8", "f8"]
-    fieldnames = ["source_id", "ra_j2000", "dec_j2000", "flux_density"]
+    fieldtypes = ["U" + str(max_name_len), "f8", "f8"]
+    fieldnames = ["source_id", "ra_j2000", "dec_j2000"]
+    stokes_names = ["I", "Q", "U", "V"]
     fieldshapes = [()] * 3
 
-    n_stokes = 4
+    n_stokes = 0
+    stokes_keep = []
+    for si, total in enumerate(np.sum(sky.stokes, axis=(1, 2))):
+        if total > 0:
+            fieldnames.append(stokes_names[si])
+            fieldshapes.append((sky.Nfreqs,))
+            fieldtypes.append("f8")
+            n_stokes += 1
+        stokes_keep.append(total > 0)
 
     if sky.freq_array is not None:
         if sky.spectral_type == "subband":
             fieldnames.append("subband_frequency")
         else:
             fieldnames.append("frequency")
         fieldtypes.append("f8")
-        fieldshapes.extend([(sky.Nfreqs, n_stokes,), (sky.Nfreqs,)])
+        fieldshapes.extend([(sky.Nfreqs,)])
     elif sky.reference_frequency is not None:
         # add frequency field (a copy of reference_frequency) for backwards
         # compatibility.
         warnings.warn(
-            "The frequency field is included in the recarray as a copy of the "
-            "reference_frequency field for backwards compatibility. In future "
-            "only the reference_frequency will be included.",
+            "The reference_frequency is aliased as `frequency` in the recarray "
+            "for backwards compatibility. In the future "
+            "only `reference_frequency` will be an accepted column key.",
             category=DeprecationWarning,
         )
-        fieldnames.extend(["frequency", "reference_frequency"])
+        fieldnames.extend([("frequency", "reference_frequency")])
         fieldtypes.extend(["f8"] * 2)
-        fieldshapes.extend([(n_stokes,)] + [()] * 2)
+        fieldshapes.extend([()] * n_stokes + [()] * 2)
         if sky.spectral_index is not None:
             fieldnames.append("spectral_index")
             fieldtypes.append("f8")
             fieldshapes.append(())
-    else:
-        # flat spectrum, no freq info
-        fieldshapes.append((n_stokes,))
```

Comment:
What happens for flat spectrum objects that don't have a reference frequency?

Reply:
This line was just adding the flux_density column to the array. Since that's handled earlier (lines 921 to 929), there's no longer anything for to do here.